### PR TITLE
Pb99u069/875 eliminate try catch

### PR DESCRIPTION
**Pull request by pb99u069**
Partially addresses #875 

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API features or modification of existing behavior is covered as defined in our [test plan](https://github.com/Plutonomicon/cardano-transaction-lib/blob/develop/doc/test-plan.md)
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:17]: This check is not enough, it does not account for overflows. Please find the boundary sizes for BigNum and ensure that the numbers are in range by converting their string representations to native JS BigInt and comparing with the boundary,

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:18]: We should not remove try/catch entirely. We want to just add more checks _before_ using try-catch. The problem with wasm-bindgen arises only when an object is allocated in a _failing_ try block.

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:24]: Leftover

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:93]: You need to find a single prefix. Construct a PublicKey value and call `to_bech32` on it to see the prefix. The simplest way is to add `console.log(key.to_bech32())` and run the test suite.

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:94]: Not `BigInt`

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:165]: Same - no overflow checks

File: src/Serialization/BigInt.js
Comment (klntsky) [inline src/Serialization/BigInt.js:13]: I think we can try constructing JS BigInt to check CSL's BigInt string for validity instead of relying on a regexp.

File: src/Types/BigNum.js
Comment (klntsky) [inline src/Types/BigNum.js:44]:
```suggestion
    check_mul(BigInt(lhs.to_str()), BigInt(rhs.to_str()), BigInt("18446744073709551616")); // 2 ^ 64
```

same above

File: src/Types/BigNum.js
Comment (klntsky) [inline src/Types/BigNum.js:17]: These two functions can be factored out into a single one. Provide the operation (`+` or `*`) as a function-parameter.

**Review by klntsky — CHANGES_REQUESTED** [review]

File: src/Serialization.js
Comment (pb99u069) [reply src/Serialization.js]: removed the regex, as `BigInt(string)` in the boundary check already throws an error if the string is incorrect

File: src/Serialization.js
Comment (pb99u069) [reply src/Serialization.js]: removed regex as well, as str is constructed from bigInt it is not necessary

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:113]: Why `pk1` and not `pk`? I'm not sure about this, have you found the code it originates from?

File: src/Serialization.js
Comment (klntsky) [inline src/Serialization.js:196]: These should be JS `BigInt`s and not JS `Number`s. In fact, they overflow here.

Comment (klntsky) [inline src/Serialization.js:196]: Also I am not sure about the exact limits, is it strictly less for positive and less-or-equal for negative? Please add the corner cases to the tests.

File: src/Serialization/BigInt.js
Comment (klntsky) [inline src/Serialization/BigInt.js]: No need for a variable.

**Review by klntsky — CHANGES_REQUESTED** [review]

File: test/Serialization.purs
Comment (klntsky) [inline test/Serialization.purs:123]:
```suggestion
        newBigNum maybeFfiHelper bn `shouldSatisfy` isJust
```

No need for parentheses here and below

**Review by klntsky — CHANGES_REQUESTED** [review]

File: src/Serialization.js
Comment (jy14898) [reply src/Serialization.js:113]: @klntsky 1 is the separator for bech32, so this is fine.

@pb99u069 The payload (hex part) must be 32 bytes long
https://github.com/Emurgo/cardano-serialization-lib/blob/04c4e7a0ee1e65253a47543be303740902820156/rust/src/chain_crypto/algorithms/ed25519.rs#L55
https://docs.rs/ed25519-dalek/0.8.0/ed25519_dalek/constant.PUBLIC_KEY_LENGTH.html

So we can also check if the string is `32/4+"ed25519_pk1".length` long

Comment (jy14898) [reply src/Serialization.js:113]: @klntsky 1 is the separator for bech32, so this is fine.

@pb99u069 The payload (hex part) must be 32 bytes long
https://github.com/Emurgo/cardano-serialization-lib/blob/04c4e7a0ee1e65253a47543be303740902820156/rust/src/chain_crypto/algorithms/ed25519.rs#L55
https://docs.rs/ed25519-dalek/0.8.0/ed25519_dalek/constant.PUBLIC_KEY_LENGTH.html

So we can also check if the string is `32*8/4+"ed25519_pk1".length` long

File: src/Serialization.js
Comment (jy14898) [inline src/Serialization.js:201]:
```suggestion
    check_limit(
      bigInt,
      BigInt("-18446744073709551615"),
      BigInt("18446744073709551616")
    ); // 2 ^ 64
    const str = bigInt.to_str();
```

File: src/Types/BigNum.js
Comment (jy14898) [inline src/Types/BigNum.js:66]: Here, the arguments to `mul` are given as strings, but `mul` is using the built-in `*` js operator.

Perhaps this fixes:
```suggestion
    check_limit_with_func(
      BigInt(lhs.to_str()),
      BigInt(rhs.to_str()),
      BigInt("18446744073709551616"),
      mul
    ); // 2 ^ 64
```

Comment (jy14898) [reply src/Types/BigNum.js:66]: Same with the `add` version. I also don't think we need this `check_limit_with_func`, just apply the function directly?

File: test/Serialization.purs
Comment (jy14898) [inline test/Serialization.purs:124]:
```suggestion
      test "BigNum overflow" $ do
```

**Review by jy14898 — CHANGES_REQUESTED** [review]

File: src/Serialization.js
Comment (ngua) [inline src/Serialization.js:14]: This applies to several other new functions in this PR: please change the names to use `camelCase` instead of `snake_case`, since that is the idiom in Javascript. CSL exports `snake_case` identifiers because the library itself is written in Rust which uses that casing convention (the JS bindings are generated from the Rust code)

File: src/Types/BigNum.js
Comment (ngua) [inline src/Types/BigNum.js:24]: See my other comments about casing

File: src/Types/BigNum.js
Comment (ngua) [inline src/Types/BigNum.js:16]: Since you're using an arrow function (which is good, since that's consistent with the rest of our JS code) and the function body is a single line, you can omit the `return` and the braces

```suggestion
const add = (a, b) => a + b;
```

File: src/Types/BigNum.js
Comment (ngua) [inline src/Types/BigNum.js:20]: Same here

File: src/Types/BigNum.js
Comment (ngua) [inline src/Types/BigNum.js:48]: What is the significance of this number? I can guess, but it should be bound to a variable with a descriptive name. The literal is repeated as well in different places, which makes all of the checks more susceptible to typos

File: src/Serialization.js
Comment (ngua) [inline src/Serialization.js:26]: See my other comment about magic numbers

**Review by ngua — CHANGES_REQUESTED** [review]

File: src/Internal/Serialization.js
Comment (ngua) [inline src/Internal/Serialization.js:128]: I'm assuming these are leftover `log`s?

File: src/Internal/Serialization.js
Comment (ngua) [inline src/Internal/Serialization.js:208]: Redundant comment

**Review by ngua — COMMENTED** [review]
Just a few final comments from me

File: src/Serialization.js
Comment (klntsky) [reply src/Serialization.js]: @jy14898 what did you mean by this suggestion? The first BigInt is CSL BigInt, and the other is native. Can't pass one instead of the other.

File: src/Serialization.js
Comment (jy14898) [reply src/Serialization.js]: The maths isn't that simple (nor is it hex)

File: src/Internal/Serialization.purs
Comment (jy14898) [inline src/Internal/Serialization.purs:22]: We should remove `newBigNum` from serialization - nothing uses it. Replace your uses in `Test.Serialization` with `Types.BigNum.fromString`

File: src/Internal/Serialization.js
Comment (jy14898) [inline src/Internal/Serialization.js:33]: See the other comment on this function. TL;DR it should be removed

**Review by jy14898 — COMMENTED** [review]
Generally LGTM. It does worry me when we have PRs that don't test the specific issue.

I did have an idea of how we can test specifically for memory leaks: allocate something (BigNum is fine) and access the internal pointer, run the test, and allocate again and check the distance between the pointers. Might be fickle though, depends on how `malloc` is implemented in CSL.

That shouldn't hold this PR up however

File: CHANGELOG.md
Comment (jy14898) [inline CHANGELOG.md:51]: Generally we want to avoid including internal changes in the changelog, otherwise it'll get cluttered and it'll be of less use to users

```suggestion
```

**Review by jy14898 — CHANGES_REQUESTED** [review]

**Review by jy14898 — APPROVED** [review]
LGTM If the merge succeeds on CI

**Pull request merged**
